src/app/services/item-set.service.ts

```diff
@@ -0,0 +1,70 @@
+import { Injectable } from '@angular/core';
+
+import { SetBonusDetail } from '~models';
+
+import { CharacterService } from './character.service';
+import { ConditionDataService } from './condition-data.service';
+import { ItemDataService } from './item-data.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ItemSetService {
+  constructor(
+    private itemDataService: ItemDataService,
+    private characterService: CharacterService,
+    private conditionDataService: ConditionDataService
+  ) {}
+
+  getSetBonusDetail(objectId: number): SetBonusDetail {
+    const result: SetBonusDetail = { pieces: [], conditions: [] };
+    const equippedItemIds = this.characterService.$character.value.inventory
+      .slice(51, 58 + 1)
+      .map(inventorySlot => inventorySlot.objectID);
+
+    // Check if item belongs to a set, if not, return empty
+    const itemData = this.itemDataService.getData(objectId);
+    if (!itemData?.setBonusId) {
+      return result;
+    }
+
+    const setBonus = this.itemDataService.setBonus[itemData.setBonusId];
+
+    // Go over all equipped pieces and check if they belong to the current setBonus
+    // These will be our highlighted (equipped items)
+    const pieces = [];
+    const tmp = [];
+    for (const equippedItemId of equippedItemIds) {
+      if (setBonus.pieces.includes(equippedItemId)) {
+        const name = this.itemDataService.getItemDetail(equippedItemId).name;
+        pieces.push({ name, isHighlighted: true });
+        tmp.push(equippedItemId);
+      }
+    }
+    // Now for the items that are not equipped, thus not highlighted
+    let difference = setBonus.pieces.filter(x => !tmp.includes(x));
+    for (let objectId of difference) {
+      const name = this.itemDataService.getItemDetail(objectId).name;
+      pieces.push({ name, isHighlighted: false });
+    }
+
+    result.pieces = [...new Map(pieces.map(item => [item['name'], item])).values()];
```

Comment:
I think this line can be simplified to `result.pieces = pieces;`

Reply:
simplyfying this line leads to redundant display of required armor pieces for a set-bonus when e. g. two of the same rings are equipped

Reply:
Please provide a comment for this line with a link to this [Stackoverflow Answer](https://stackoverflow.com/a/58429784/11125147)